### ISSUE TITLE
Wiki.java TODO/whinge list and thoughts

### ISSUE THREAD

**Issue by MER-C** [MER-C/wiki-java#154]
## Vectorization

* [ ] Revision.getText: can be merged into getPageText(String[]). Will need some fiddling around with the internals, I get the feeling that constructRevisionString and constructTitleString can be merged once the JDK is up to scratch.
* [ ] getLastRevision
* [ ] getFirstRevision: https://phabricator.wikimedia.org/T188672

## MediaWiki annoyances and wishlist

* [ ] Revisions from contribs(X) don't have SHA-1 populated: https://phabricator.wikimedia.org/T185809
* [ ] Revisions from watchlist don't have SHA-1 populated: https://phabricator.wikimedia.org/T185808
* [ ] Range contribs through the API: https://phabricator.wikimedia.org/T177150. Remove rangeContribs once this is done.
* [ ] getBlockList() returning rangeblocks if given a specific IP address: https://phabricator.wikimedia.org/T183300
* [ ] No [[Special:Recentchangeslinked]]: https://phabricator.wikimedia.org/T17552

## General FIXMEs
* [ ] parse+diff: permissiondenied and missingtitle are generic error messages that represent real unrecoverable problems in other methods.

## Blocked by JDK availability on Google App Engine
Current GAE stable JDK version is 8.
* [ ] todo

## Blocked by other
* [ ] Move unit tests to JUnit Jupiter (the new syntax, especially for testing ## exceptions). Blocked by Netbeans.

## Tools
* [ ] CSV export (revisions, log entries)

**Comment by aTOmQpLUs** [comment]
🙈🙉🙊🔊🕸🏆

**Comment by aTOmQpLUs** [comment]
🙈ftp.🙉php.🙊pdf.atomxplus.com

**Comment by Abbe98** [comment]
I was wondering if anyone else got a use case for having the two upload methods support warnings. Turning `ignorewarnings` into a method parameter and letting the methods return a list of warnings if it's not enabled. 

https://www.mediawiki.org/wiki/API:Upload#Upload_warnings

This would be very useful for Pattypan as it could serve as a neat solution to detect duplicates and raise other potential issues.

**Comment by MER-C** [comment]
I did the obvious thing with fd9b2dd5.

**Comment by Abbe98** [comment]
I'm afraid https://github.com/MER-C/wiki-java/commit/fd9b2dd51623220a197f5829c05bb4ada4757e72 won't solve it on our end as we want to detect duplicates prior to or during the upload. We can solve it by just sending the hash to the WM-API but I imagine more upload tools could benefit for a way to get upload warnings.
